### PR TITLE
[Client] Add metadata to Terminate Calls to make ray.kill() and ray.cancel() work

### LINKED ISSUE #15175: [ray client] Support public API: ray.cancel()

### LINKED ISSUE #15176: [ray client] Support public API: ray.kill()

### PR DESCRIPTION
**Pull request by ijrsvt**
## Why are these changes needed?
* Metadata is not set in these calls, meaning that issues may arise when making this calls!

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by AmeerHajAli — COMMENTED** [review]
Can we add a unit test?

**Comment by DmitriGekhtman** [comment]
> Can we add a unit test?

Yeah, I can take care of that.

**Comment by DmitriGekhtman** [comment]
The bug here was that `metadata` was not supplied to a method of the stub `self.server`.
Shall we wrap the stub to make `metadata` a required arg of all of its methods? What do you guys think?

**Comment by ijrsvt** [comment]
> Shall we wrap the stub to make `metadata` a required arg of all of its methods? What do you guys think?

That honestly makes a lot of sense!

**Comment by AmeerHajAli** [comment]
> The bug here was that `metadata` was not supplied to a method of the grpc stub `self.server`.
> Shall we wrap the stub to make `metadata` a required arg of all of its methods? What do you guys think?

+1 I like the proposal. But please make sure that it can be set to None so that users with metadata as None (default) should be allowed to connect.

**Comment by DmitriGekhtman** [comment]
> make sure that it can be set to None so that users with metadata as None (default) should be allowed to connect

This should work out -- at the point where the methods of the stub are being called, the default of None for metadata has already been converted to [] .

**Comment by AmeerHajAli** [comment]
> > make sure that it can be set to None so that users with metadata as None (default) should be allowed to connect
> 
> This should work out -- at the point where the methods of the stub are being called, the default of None for metadata has already been converted to [] .

We should not have default values of mutable objects

**Comment by DmitriGekhtman** [comment]
> We should not have default values of mutable objects

Ah, right, sry -- I was referring to this line:
https://github.com/ray-project/ray/blob/6709560ef6b9cee929a8af2e8186918d6fcacc0c/python/ray/util/client/worker.py#L79

**Comment by AmeerHajAli** [comment]
> The bug here was that `metadata` was not supplied to a method of the grpc stub `self.server`.
> Shall we wrap the stub to make `metadata` a required arg of all of its methods? What do you guys think?

lets add this as quickly as possible + a unit test(s).

**Comment by DmitriGekhtman** [comment]
Looked into it a bit, but realized I'm actually unsure how to unit-test this.

**Comment by DmitriGekhtman** [comment]
Will just unit test the fix by with a relevant mock.

**Comment by DmitriGekhtman** [comment]
added a test validating that metadata is being passed in the right place

**Comment by AmeerHajAli** [comment]
@DmitriGekhtman , Didn't we plan to enforce this at the top level?

**Comment by DmitriGekhtman** [comment]
> @DmitriGekhtman , Didn't we plan to enforce this at the top level?

I'll go ahead and do that.

File: python/ray/util/client/worker.py [hunk 1]
```diff
@@ -110,8 +112,9 @@ def __init__(self,
                 grpc.channel_ready_future(self.channel).result(timeout=timeout)
                 # The HTTP2 channel is ready. Wrap the channel with the
                 # RayletDriverStub, allowing for unary requests.
-                self.server = ray_client_pb2_grpc.RayletDriverStub(
-                    self.channel)
+                server = ray_client_pb2_grpc.RayletDriverStub(self.channel)
+                # Require metadata with all requests
+                self.server = require_metadata(server)
```
Comment (DmitriGekhtman) [inline python/ray/util/client/worker.py:117]: Found it easiest to just do this at runtime.
Added a test for `require_metadata`.

File: python/ray/util/client/worker.py [hunk 2]
```diff
@@ -148,6 +151,8 @@ def __init__(self,
         self.log_client.set_logstream_level(logging.INFO)
 
         self.closed = False
+        import pdb
+        pdb.set_trace()
```
Comment (ijrsvt) [inline python/ray/util/client/worker.py:155]: Can you remove this?

File: python/ray/util/client/worker.py (hunk 1 above)
Comment (AmeerHajAli) [reply python/ray/util/client/worker.py:117]: Can't we just make this a testing where we go over all the functions and verify that we can pass metadata to them? 
I find the current code a little bit hard to read and can leave hidden bugs.

File: python/ray/util/client/worker.py (hunk 2 above)
Comment (DmitriGekhtman) [reply python/ray/util/client/worker.py:155]: oops

File: python/ray/util/client/worker.py (hunk 1 above)
Comment (DmitriGekhtman) [reply python/ray/util/client/worker.py:117]: The issue was that it was possible to not pass metadata.
I think we can just remove this logic and we'll be fine.

Can you explain more precisely what you had in mind for testing?

File: python/ray/tests/test_client.py [hunk 3]
```diff
@@ -493,5 +494,21 @@ def stop_server(server):
...
+
+    stub = MockDriverStub()
+    # Metadata not required in the Terminate method
+    assert stub.Terminate("foo") is None
+    modified_stub = require_metadata(stub)
+    # If metadata is provided, things go as expected.
+    assert modified_stub.Terminate("foo", metadata="bar") == "bar"
+    # If not, there's a TypeError.
+    with pytest.raises(TypeError):
+        modified_stub.Terminate("foo")
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client.py:510]: Can we verify `modified_stub.Terminate("foo", None)` works here as well?

File: python/ray/util/client/worker.py (hunk 1 above)
Comment (DmitriGekhtman) [reply python/ray/util/client/worker.py:117]: This is indeed hard to read and prone to bugs.
Reverting.

**Comment by DmitriGekhtman** [comment]
> Shall we wrap the stub to make `metadata` a required arg of all of its methods? 

This complicated things needlessly. Have reverted the change implementing that.

Should be fine for a final review now.

File: python/ray/tests/test_client_terminate.py [hunk 4]
```diff
@@ -88,6 +89,39 @@ def wait_for(t):
         ray.get(obj1)
 
 
+def test_kill_cancel_metadata(ray_start_regular):
+    """
+    Verify that client worker's terminate_actor and terminate_task methods
+    pass worker's metadata attribute server to the grpc stub's Terminate
+    method.
+    """
+    with ray_start_client_server(metadata=[("foo", "bar")]) as ray:
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:98]:
```suggestion
    with ray_start_client_server(metadata=[("key", "value")]) as ray:
```

File: python/ray/tests/test_client_terminate.py [hunk 5]
```diff
@@ -88,6 +89,39 @@ def wait_for(t):
         ray.get(obj1)
 
 
+def test_kill_cancel_metadata(ray_start_regular):
+    """
+    Verify that client worker's terminate_actor and terminate_task methods
+    pass worker's metadata attribute server to the grpc stub's Terminate
+    method.
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:96]:
```suggestion
    method. This is done by raising an exception with contents of metadata
    and verifying the exception is raised.
```

File: python/ray/tests/test_client_terminate.py [hunk 6]
```diff
@@ -88,6 +89,39 @@ def wait_for(t):
...
+
+        @ray.remote
+        def f():
+            time.sleep(1000)
+
+        actor = A.remote()
+        task_ref = f.remote()
+
+        def mock_terminate(term, metadata):
+            raise Exception(metadata[0][0])
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:112]:
```suggestion
            raise MetaDataIsCorrectlyPassedException(metadata[0][0])
```

File: python/ray/tests/test_client_terminate.py [hunk 7]
```diff
@@ -88,6 +89,39 @@ def wait_for(t):
...
+
+        ray.api.worker.server.Terminate = mock_terminate
+
+        with pytest.raises(Exception) as e:
+            ray.kill(actor)
+        assert str(e.value) == "foo"
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:118]:
```suggestion
            assert str(e.value) == "foo"
```

File: python/ray/tests/test_client_terminate.py [hunk 8]
```diff
@@ -88,6 +89,39 @@ def wait_for(t):
...
+
+        with pytest.raises(Exception) as e:
+            ray.cancel(task_ref)
+        assert str(e.value) == "foo"
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:122]:
```suggestion
            assert str(e.value) == "foo"
```

File: python/ray/tests/test_client_terminate.py (hunk 6 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_client_terminate.py:112]: Raising an exception here is properly not the right idiom at all.
I think `unittest` provides a framework for side effects to do this. Will take a brief read.

Comment (DmitriGekhtman) [reply python/ray/tests/test_client_terminate.py:112]: eh, too lazy, will define a more descriptive Exception class, as you've suggested

File: python/ray/tests/test_client_terminate.py (hunk 7 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_client_terminate.py:118]: This isn't correct -- if the indent is added, the assert statement will not be executed.

File: python/ray/tests/test_client_terminate.py [hunk 9]
```diff
@@ -105,21 +109,26 @@ class A:
...
-            raise Exception(metadata[0][0])
+            raise MetadataIsCorrectlyPassedException(metadata[0][0])
 
+        # Mock stub's Terminate method to raise an exception.
         ray.api.worker.server.Terminate = mock_terminate
 
-        with pytest.raises(Exception) as e:
+        # Verify the expected exception is raised with ray.kill.
+        with pytest.raises(MetadataIsCorrectlyPassedException) as e:
+            actor = A.remote()
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:123]: please remove this outside the with statement.

File: python/ray/tests/test_client_terminate.py [hunk 10]
```diff
@@ -105,21 +109,26 @@ class A:
...
             ray.kill(actor)
-        assert str(e.value) == "foo"
+        assert str(e.value) == "key"
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:125]: can you please add a tab before this to be below the with statement?

File: python/ray/tests/test_client_terminate.py [hunk 11]
```diff
@@ -105,21 +109,26 @@ class A:
...
 
-        with pytest.raises(Exception) as e:
+        # Verify the expected exception is raised with ray.cancel.
+        with pytest.raises(MetadataIsCorrectlyPassedException) as e:
+            task_ref = f.remote()
             ray.cancel(task_ref)
-        assert str(e.value) == "foo"
+        assert str(e.value) == "key"
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:131]: can you please add a tab before this to be below the with statement?

File: python/ray/tests/test_client_terminate.py [hunk 12]
```diff
@@ -105,21 +109,26 @@ class A:
...
+        with pytest.raises(MetadataIsCorrectlyPassedException) as e:
+            actor = A.remote()
             ray.kill(actor)
-        assert str(e.value) == "foo"
+        assert str(e.value) == "key"
 
-        with pytest.raises(Exception) as e:
+        # Verify the expected exception is raised with ray.cancel.
+        with pytest.raises(MetadataIsCorrectlyPassedException) as e:
+            task_ref = f.remote()
```
Comment (AmeerHajAli) [inline python/ray/tests/test_client_terminate.py:129]: please remove this outside the with statement.

File: python/ray/tests/test_client_terminate.py (hunk 11 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_client_terminate.py:131]: nack
If I add a tab, the assert statement will not be executed.

Comment (DmitriGekhtman) [reply python/ray/tests/test_client_terminate.py:131]: just took another look at pytest docs

there's a way of making this cleaner which I will add in a moment

File: python/ray/tests/test_client_terminate.py (hunk 9 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_client_terminate.py:125]: sure

File: python/ray/tests/test_client_terminate.py (hunk 12 above)
Comment (DmitriGekhtman) [reply python/ray/tests/test_client_terminate.py]: sure

**Review by AmeerHajAli — APPROVED** [review]

**Pull request merged**